### PR TITLE
Use `helpers.shema.Provisoner` in chef provisioner

### PR DESCRIPTION
**Pull request by VladRassokhin**
Also `ValidateFunc` introduced for extended provisioner data validation.

This PR is rebase of #10422 on top of master after #10934 being merged.

### CODE REVIEW

**Comment by VladRassokhin** [comment]
@mitchellh Please take a look. I've rebased my branch after your changes in provisioners

File: builtin/provisioners/chef/resource_provisioner.go [hunk 1]
```diff
@@ -727,3 +825,57 @@ func retryFunc(timeout time.Duration, f func() error) error {
...
+		DisableReporting:      d.Get("disable_reporting").(bool),
+		Environment:           d.Get("environment").(string),
+		FetchChefCertificates: d.Get("fetch_chef_certificates").(bool),
+		LogToFile:             d.Get("log_to_file").(bool),
+		UsePolicyfile:         d.Get("use_policyfile").(bool),
+		PolicyGroup:           d.Get("policy_group").(string),
+		PolicyName:            d.Get("policy_name").(string),
+		HTTPProxy:             d.Get("http_proxy").(string),
+		HTTPSProxy:            d.Get("https_proxy").(string),
+		NOProxy:               getStringList(d.Get("no_proxy")),
```
Comment (VladRassokhin) [inline builtin/provisioners/chef/resource_provisioner.go]: NameRunList is missing here

**Comment by VladRassokhin** [comment]
@paddycarver @mitchellh  Sorry for disturbing but could anyone merge/comment on this PR ❓

**Comment by svanharmelen** [comment]
@VladRassokhin I will review this one, but it will take a few more days until I am behind a desk again (and this PR is a bit too big to review on my phone :wink:) So unless someone else steps in, I'll make sure to make some time for this one in the coming week.

**Comment by svanharmelen** [comment]
@VladRassokhin I spent a reasonable amount of time on this one today... I started with a normal review, but it soon became a challenge to mark and explain my findings/concerns.

So I thought the fasted way forward was to created a new branch based on your work and add a single commit tweaking what was (IMO) needed to get this ready to be merged.

Overall I didn't really change any fundamental stuff. There was one real issue with the schema (the fields that contained an `Elem`) and other then that I just did some refactoring and cleaning (removed the deprecated fields).

So thank you very much for your work on this one! And my apologies for the leak of response on this one! I'll be closing this PR as soon as the new one is merged (which will hopefully be today).

**Comment by svanharmelen** [comment]
OK, so I'm going to close this one right away as I need an additional `LGTM` on the new PR since I now added code myself as well :wink:

So to not cause any confusion with these almost similar PR's, I'll close this one and refer you to the new PR #14681.

Thanks again!!

**Pull request closed without merging**